### PR TITLE
fix: update GW cache outside LN module init

### PR DESCRIPTION
**Pull request by shaurya947**
Phase 1 of the 3-phase fix described in #6153 .. other phases will be their own separate PRs as and when those approaches are aligned on

### CODE REVIEW

File: modules/fedimint-ln-client/src/lib.rs [hunk 1]
```diff
@@ -618,16 +618,8 @@ impl LightningClientModule {
             secp,
             client_ctx: args.context(),
             update_gateway_cache_merge: UpdateMerge::default(),
-            gateway_conn: gateway_conn.clone(),
-        };
-
-        // Only initialize the gateway cache if it is empty
```
Comment (dpc) [inline modules/fedimint-ln-client/src/lib.rs:624]: Not doing it here is a significant change API change. I don't have a full understanding of the consequences.

But generally yes, this needs to be fixed, and doing rpc calls here slows the client initialization, and possibility of failing whole client due to a connectivity problem is terrible.

Alternatively, we should spawn a background task or something to keep doing it. Though the downside is that it racy and non-deterministic, so I'm a bit meh on it.

**Comment by dpc** [comment]
@fedimint/lightning should look and discuss, but probably OK

**Comment by Kodylow** [comment]
maanmeet says its okay

File: fedimint-load-test-tool/src/main.rs [hunk 2]
```diff
@@ -534,6 +534,10 @@ async fn get_user_client(
         invite_code.clone()
     };
     let (client, invite_code) = build_client(user_invite_code, user_db.as_ref()).await?;
+    // if lightning module is present, update the gateway cache
+    if let Ok(ln_client) = client.get_first_module::<LightningClientModule>() {
+        let _ = ln_client.update_gateway_cache().await;
```
Comment (dpc) [inline fedimint-load-test-tool/src/main.rs:539]: Is ignoring the error potential source of flakiness?

File: modules/fedimint-ln-tests/tests/tests.rs [hunk 3]
```diff
@@ -48,6 +48,7 @@ async fn pay_invoice(
     gateway_id: Option<secp256k1::PublicKey>,
 ) -> anyhow::Result<OutgoingLightningPayment> {
     let ln_module = client.get_first_module::<LightningClientModule>()?;
+    ln_module.update_gateway_cache().await?;
```
Comment (dpc) [inline modules/fedimint-ln-tests/tests/tests.rs:51]: And here we don't ignore?

**Comment by shaurya947** [comment]
I ignored in places where lnv2 is also present. Lnv2 doesn’t need this
additional step to update cache right after client init. So I figured when
those tests deprecate lnv1 all this code would be removed anyway.

However I did not ignore the res value in tests that are specific to lnv1

**Comment by dpc** [comment]
I bet lnv1 will stick around for a while. 


I haven't thought much about it, but it seems to me we need to handle possibility of a network failure, so neither ignoring nor failing are a good strategy. I think it would be more robust to have some kind of a retry loop function there and use that everywhere.

`await_gateway_cache_populated().await` or something?

**Comment by shaurya947** [comment]
I don’t think retry logic is necessary since master doesn’t currently have
it (in lightning init) and AFAIK it hasn’t resulted in flakiness. But if we
want full code parity perhaps we should fail if the result is an err
everywhere

**Comment by dpc** [comment]
I'm thinking more about the desired state than whatever is there now. Might as well steer everything in desired direction while at it.

I am guessing that test will want to make gateway cache to get populated at least once somewhere early, while  end user apps spawn a task that will keep populating it periodically(?). So we should probably have these right away.

**Comment by shaurya947** [comment]
I think you are correct: tests will want to update cache once during setup,
whereas client integrators will want to call the function
update_gateway_cache_continously (might be remembering the name wrong but
it’s there in LightningClientModule).

In this PR, tests already take that approach. For fedimint-cli IDK if it
makes sense to call the “continuously” function. What do you think? We can
also iterate in follow up PRs if we want to deliberate more on the
fine-grained approach.

**Review by dpc — APPROVED** [review]

**Comment by dpc** [comment]
We'll figure it out. :)

**Pull request closed without merging**